File: python/1-two-sum.py
from typing import List
from operator import itemgetter

class Solution:
    # def twoSum(self, nums: List[int], target: int) -> List[int]:
    #     "Brute Force, Time: O(n^2), Space: O(1)"
    #     n = len(nums)
    #     for i in range(n-1):
    #         for j in range(i+1, n):
    #             if nums[i] + nums[j] == target:
    #                 return [i, j]
    #     return []
    # def twoSum(self, nums: List[int], target: int) -> List[int]:
    #     "Sort + 2 Pointer, Time: O(nlogn)"
    #     indices, sorted_nums = zip(*sorted(enumerate(nums), key=itemgetter(1)))
    #     i, j = 0, len(nums)-1
    #     while i < j:
    #         if sorted_nums[i] + sorted_nums[j] == target:
    #             return [indices[i], indices[j]]
    #         elif sorted_nums[i] + sorted_nums[j] < target:
    #             i += 1
    #         else:
    #             j -= 1
    # def twoSum(self, nums: List[int], target: int) -> List[int]:
    #     "Two Pass, Time: O(n), Space: O(n)"
    #     table = {num: i for i, num in enumerate(nums)}
    #     for i, num in enumerate(nums):
    #         complement = target - num
    #         if complement in table and i != table[complement]:
    #             return [i, table[complement]]
    #     return []
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        "One Pass, Time: O(n), Space: O(n)"
        table = {}
        for i, num in enumerate(nums):
            complement = target - num
            if complement in table:
                return [i, table[complement]]
            table[num] = i
        return []

nums = [2,7,11,15]
target = 9
result = Solution().twoSum(nums, target)
assert sorted(result) == [0,1]

nums = [3,2,4]
target = 6
result = Solution().twoSum(nums, target)
assert sorted(result) == [1, 2]

nums = [3,3]
target = 6
result = Solution().twoSum(nums, target)
assert sorted(result) == [0, 1]

File: cpp/1. Two Sum.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int, int> pi;

class Solution {
// HashMap, Time: O(n), Space: O(n)
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        int n = nums.size();
        unordered_map<int, int> table; // val: index
        vi ans(2);
        for(int i=0; i<n; i++) {
            int num = nums[i];
            int remain = target - num;
            auto iter = table.find(remain);
            if(iter!=table.end()) {
                auto [val, idx] = *iter;
                ans[0] = idx;
                ans[1] = i;
            } else {
                table[num] = i;
            }
        }
        return ans;
    }
};

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

    Solution solution; vi nums; int target; vi ans;

    solution = Solution();
    nums = {2,7,11,15}, target = 9;
    ans = solution.twoSum(nums, target);
    assert(ans==vi({0, 1}));

    solution = Solution();
    nums = {3,2,4}, target = 6;
    ans = solution.twoSum(nums, target);
    assert(ans==vi({1, 2}));
}